portfolio/src/main/webapp/js/background.js

```diff
@@ -0,0 +1,291 @@
+/**
+ * @fileoverview Creates a live visualization of balls that repel each other
+ * and bounce off of the walls.
+ */
+const HEIGHT_TO_WIDTH = 0.3;
+const EFFECT_RADIUS_TO_WIDTH = 0.1;
+const BALL_SIZE_TO_WIDTH = 0.01;
+const NUM_BALLS = 25;
+const INITIAL_BALL_VELOCITY_MAGNITUDE = 1;
+/**
+ * Runs initial set up of the background and background canvas.
+ */
+function setup() {
+  const backgroundContainer = document.getElementById("background-container");
+
+  const backgroundCanvas =
+    createCanvas(backgroundContainer.clientWidth,
+      backgroundContainer.clientWidth * HEIGHT_TO_WIDTH);
+  backgroundCanvas.parent("background-container");
+
+  frameRate(60);
+
+  initTank();
+}
+
+let ballTank;
+/**
+ * Initializes a tank for the background.
+ */
+function initTank() {
+  const ballSize = BALL_SIZE_TO_WIDTH * width;
+  ballTank = new Tank(width, height, EFFECT_RADIUS_TO_WIDTH * width, ballSize);
+
+  // Create balls and add them to the tank.
+  // Each start with random positions and velocities.
+  for (let i = 0; i < NUM_BALLS; i++) {
+    const randomPos = createVector(Math.random() * ballTank.width,
+      Math.random() * ballTank.height);
+
+    // Gets a random vector on the unit circle as to randomize 
+    // the initial velocity direction.
+    const randomVelocity = p5.Vector.random2D();
+    randomVelocity.mult(INITIAL_BALL_VELOCITY_MAGNITUDE);
+
+    const newBall = new Ball(randomPos, randomVelocity);
+    ballTank.addBall(newBall);
+  }
+}
+
+/**
+ * Resizes the canvas when the window is resized.
+ */
+function windowResized() {
+  const canvasContainer = document.getElementById("background-container");
+  resizeCanvas(canvasContainer.clientWidth,
+    canvasContainer.clientWidth * HEIGHT_TO_WIDTH);
+}
+
+/**
+ * Called every draw loop; used for updating game logic
+ * and drawing to the screen.
+ */
+function draw() {
+  const grayScaleColor = 255; // white
+  background(grayScaleColor);
+  ballTank.update();
+  ballTank.draw();
+}
+
+/**
+ * Tank that holds, updates, and draws balls that repel each other.
+ */
+class Tank {
+  /**
+   * @param {number} width
+   * @param {number} height
+   * @param {number} effectRadius
+   * @param {number} ballSize
+   */
+  constructor(width, height, effectRadius, ballSize) {
+    /**
+     * Height of the tank.
+     * @type {number} @const
+     */
+    this.height = height;
+
+    /**
+     * Width of the tank.
+     * @type {number} @const
+     */
+    this.width = width;
+
+    /**
+     * Effective radius of the repulsion effect.
+     * @type {number} @const
+     */
+    this.effectRadius = effectRadius;
+
+    /**
+     * Effective diameter of the repulsion effect.
+     * @type {number} @const
+     */
+    this.effectDiameter = effectRadius * 2;
+
+    /**
+     * The size of the repelling balls.
+     * @type {number} @const
+     */
+    this.ballSize = ballSize;
+
+    /**
```

Comment:
should we remove this now that the optimization is also gone?

Reply:
This is used to store the visual size of the balls on the screen. It is unrelated to the optimization code, but if that seems unclear could you elaborate on how I could improve that?

Reply:
Hm wow the comment kind of moves around as the code changes I guess? Will have to keep that in mind. This was originally supposed to be on the lines below re: this._spatialHash.

Reply:
Oh okay that makes more sense haha. Another good catch. I'll have to avoid fixing code too late at night next time haha.